ReactiveCocoa/Swift/Flatten.swift

```diff
@@ -0,0 +1,571 @@
+	@warn_unused_result(message="Did you forget to call `start` on the producer?")
+	private func merge() -> Signal<Value.Value, Error> {
+		return Signal<Value.Value, Error> { relayObserver in
+			self.observeMerge(MergeState(observer: relayObserver, disposable: nil))
+		}
+	}
+
+	private func observeMerge(state: MergeState<Value.Value, Error>) -> Disposable? {
+		return self.observe { event in
+			switch event {
+			case let .Next(producer):
+				producer.startWithSignal { innerSignal, innerDisposable in
+					state.inFlight.modify { $0 + 1 }
+					let handle = state.disposable?.addDisposable(innerDisposable) ?? nil
+
+					innerSignal.observe { event in
+						switch event {
+						case .Completed, .Interrupted:
+							handle?.remove()
+							state.decrementInFlight()
+
+						default:
+							state.observer.action(event)
+						}
+					}
+				}
+
+			case let .Failed(error):
+				state.observer.sendFailed(error)
+
+			case .Completed:
+				state.decrementInFlight()
+
+			case .Interrupted:
+				state.observer.sendInterrupted()
+			}
+		}
+	}
+}
+
+extension SignalProducerType where Value: SignalProducerType, Error == Value.Error {
+	/// Merges a `signal` of SignalProducers down into a single signal, biased toward the producer
+	/// added earlier. Returns a Signal that will forward events from the inner producers as they arrive.
+	@warn_unused_result(message="Did you forget to call `start` on the producer?")
+	private func merge() -> SignalProducer<Value.Value, Error> {
+		return SignalProducer<Value.Value, Error> { relayObserver, disposable in
+			let state = MergeState(observer: relayObserver, disposable: disposable)
+
+			self.startWithSignal { signal, signalDisposable in
+				disposable.addDisposable(signalDisposable)
+
+				signal.observeMerge(state)
+			}
+
+		}
+	}
+}
+
+extension SignalType {
+	/// Merges the given signals into a single `Signal` that will emit all values
+	/// from each of them, and complete when all of them have completed.
+	public static func merge<S: SequenceType where S.Generator.Element == Signal<Value, Error>>(signals: S) -> Signal<Value, Error> {
+		let producer = SignalProducer<Signal<Value, Error>, Error>(values: signals)
+		var result: Signal<Value, Error>!
+
+		producer.startWithSignal { (signal, _) in
+			result = signal.flatten(.Merge)
+		}
+
+		return result
+	}
+}
+
+public final class MergeState<Value, Error: ErrorType> {
+	/// The observer of a started `concat` producer.
+	let observer: Observer<Value, Error>
+
+	/// The top level disposable of a started `concat` producer.
+	let disposable: CompositeDisposable?
+
+	/// The number of outstanding merged signals
+	let inFlight = Atomic(1)
+
+	init(observer: Observer<Value, Error>, disposable: CompositeDisposable?) {
+		self.observer = observer
+		self.disposable = disposable
+	}
+
+	func decrementInFlight() {
+		let orig = inFlight.modify { $0 - 1 }
+		if orig == 1 {
+			observer.sendCompleted()
+		}
+	}
+}
```

Comment:
I don't think `MergeState` buys us very much in terms of abstraction. IMO it makes things a little harder to read since the `sendCompleted` in `decrementInFlight` is the only real logic. I would create and use the `Atomic` directly from `observeMerge` and get rid of this class.